### PR TITLE
Update example to use Ray dataset and Ray train

### PR DESCRIPTION
**Pull request by KepingYan**

### CODE REVIEW

File: examples/raysgd_nyctaxi.py [hunk 1]
```diff
@@ -61,7 +70,7 @@ def __init__(self, cols):
         self.bn4 = nn.BatchNorm1d(16)
 
     def forward(self, *x):
-        x = torch.cat(x, dim=1)
+        x = x[0]
```
Comment (kira-lin) [inline examples/raysgd_nyctaxi.py:73]: what was the data shape? does this lose input data?

File: examples/raysgd_nyctaxi.py [hunk 2]
```diff
@@ -84,42 +93,66 @@ def setup(self, config):
...
+
+def train_func(config):
+    device = torch.device(f"cuda:{train.local_rank()}" if
+                          torch.cuda.is_available() else "cpu")
+    if torch.cuda.is_available():
+        torch.cuda.set_device(device)
+
+    # Set the args to whatever values you want.
+	# If using DistributedSampler, modify it here.
+    training_operator = CustomOperator(
```
Comment (kira-lin) [inline examples/raysgd_nyctaxi.py:110]: With ray train, we no longer need this operator, right? Let's use pure ray train(which is the new version of raysgd), check their [examples](https://docs.ray.io/en/latest/train/examples.html).

File: examples/raysgd_nyctaxi.py [hunk 3]
```diff
@@ -38,12 +43,16 @@
...
+train_dataset = ray.data.from_spark(train_df, parallelism = num_executors)
+test_dataset = ray.data.from_spark(test_df, parallelism = num_executors)
 # Then convert to torch datasets
-train_dataset = train_dataset.to_torch(feature_columns=features, label_column="fare_amount")
-test_dataset = test_dataset.to_torch(feature_columns=features, label_column="fare_amount")
+feature_dtype = [torch.float] * len(features)
+train_dataset = train_dataset.to_torch(feature_columns=features, label_column="fare_amount",
+                                               label_column_dtype=torch.float, feature_column_dtypes=feature_dtype, batch_size=64)
+test_dataset = test_dataset.to_torch(feature_columns=features, label_column="fare_amount",
+                                             label_column_dtype=torch.float, feature_column_dtypes=feature_dtype, batch_size=64)
```
Comment (kira-lin) [inline examples/raysgd_nyctaxi.py:55]: We should get shards of the dataset inside the train_func, check [here](https://docs.ray.io/en/latest/train/user_guide.html#train-datasets).

**Review by kira-lin — COMMENTED** [review]
We should also rename it to raytrain_nyctaxi.py.

File: examples/raysgd_nyctaxi.py (hunk 1 above)
Comment (KepingYan) [reply examples/raysgd_nyctaxi.py]: I modified parameter "*x" -> "x", then the type of x is `torch.Tensor` and the shape is `[batchsize, len(features)]`.

File: examples/raytrain_nyctaxi.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
+    loss_results = []
+    for epoch in range(num_epochs):
+        train_acc, train_loss = train_epoch(train_dataset, model, criterion, optimizer, device)
+        test_acc, test_loss = test_epoch(test_dataset, model, criterion, device)
+        train.report(epoch = epoch, train_acc = train_acc, train_loss = train_loss)
+        train.report(epoch = epoch, test_acc=test_acc, test_loss=test_loss)
+        loss_results.append(test_loss)
+
+trainer = Trainer(backend="torch", num_workers=num_executors, use_gpu=False)
```
Comment (kira-lin) [inline examples/raytrain_nyctaxi.py:151]: Let num_workers be 2 to make it distributed. Claim num_cpus=4 in ray.init to pass CI.
@carsonwang Shall we include gpu support? We can't test it. This comes from a ray train example. If we support gpu, use_gpu should be a config, refer to the horovod example about how to set args.

File: examples/raytrain_nyctaxi.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+# Set spark timezone for processing datetime
+spark.conf.set("spark.sql.session.timeZone", "UTC")
+# Transform the dataset
+data = nyc_taxi_preprocess(data)
+# Split data into train_dataset and test_dataset
+train_df, test_df = random_split(data, [0.9, 0.1], 0)
+features = [field.name for field in list(train_df.schema) if field.name != "fare_amount"]
+
+# Convert spark dataframe into ray Dataset
+train_dataset = ray.data.from_spark(train_df, parallelism = num_executors)
```
Comment (kira-lin) [inline examples/raytrain_nyctaxi.py:50]: Remember to align parallelism with num_workers of ray train.

File: examples/raytrain_nyctaxi.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+def train_epoch(dataset, model, criterion, optimizer, device):
+    model.train()
+    train_loss, correct, data_size = 0, 0, 0
+    for batch_idx, (inputs, targets) in enumerate(dataset):
+        inputs, targets = inputs.to(device), targets.to(device)
+        # Compute prediction error
+        outputs = model(inputs)
+        loss = criterion(outputs, targets)
+        train_loss += loss.item()
+        correct += (outputs.argmax(1) == targets).type(torch.float).sum().item()
```
Comment (kira-lin) [inline examples/raytrain_nyctaxi.py:96]: correct is not calculated as this. This only applies to the mnist example of ray. We should compare outputs to targets to calculate accuracy.

File: examples/raytrain_nyctaxi.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    num_epochs = config["num_epochs"]
+    lr = config["lr"]
+    batch_size = config["batch_size"]
+    device = torch.device(f"cuda:{train.local_rank()}"
+                          if torch.cuda.is_available() else "cpu")
+    
+    # Then convert to torch datasets                
+    train_data_shard = get_dataset_shard("train")
+    train_dataset = train_data_shard.to_torch(feature_columns=features, label_column="fare_amount",
+                                               label_column_dtype=torch.float, feature_column_dtypes=feature_dtype, batch_size=batch_size)
```
Comment (kira-lin) [inline examples/raytrain_nyctaxi.py:134]: This line might be too long

**Review by kira-lin — COMMENTED** [review]
Almost LGTM, but it seems like we have some lint error. Can you add this examples folder to the CI lint check as well? If there are lint errors in other existing examples, fix them too.

File: .github/workflows/raydp.yml [hunk 8]
```diff
@@ -94,6 +94,7 @@ jobs:
         run: |
           pip install pylint==2.8.3
           pylint --rcfile=python/pylintrc python/raydp
+          pylint --rcfile=python/pylintrc examples/raytrain_nyctaxi.py
```
Comment (kira-lin) [inline .github/workflows/raydp.yml:97]: make it examples/*.py?

Comment (KepingYan) [reply .github/workflows/raydp.yml:97]: Ok, I'll push again after modifying other files.

**Comment by carsonwang** [comment]
@KepingYan , are those formatting changes automatically from your IDE settings? If you do want to update them, we'd better to do them in a separate PR.

**Comment by KepingYan** [comment]
changes based on previous test results:

- transfer the train, test dataset split operation on ray dataset.

- during dataframe->ray data, use ‘repartition()’ to perform resharding operation.

- switch evaluation metric from accuracy to MSE. The effect can be seen on source dataset, and is not obvious on randomly obtained dataset.

File: examples/raytrain_nyctaxi.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+data = spark.read.format("csv").option("header", "true") \
+        .option("inferSchema", "true") \
+        .load(NYC_TRAIN_CSV)
+# Set spark timezone for processing datetime
+spark.conf.set("spark.sql.session.timeZone", "UTC")
+# Transform the dataset
+data = nyc_taxi_preprocess(data)
+# Convert spark dataframe into ray Dataset
+# pylint: disable=unbalanced-tuple-unpacking
+dataset = ray.data.from_spark(data, parallelism = num_executors)
```
Comment (carsonwang) [inline examples/raytrain_nyctaxi.py:43]: Remove parallelism here?

Comment (KepingYan) [reply examples/raytrain_nyctaxi.py]: Done.
